google/cloud/forseti/scanner/audit/retention_rules_engine.py

```diff
@@ -0,0 +1,327 @@
+            if resource_ids is None:
+                raise audit_errors.InvalidRulesSchemaError(
+                    'Lack of resource_ids in rule {}'.format(rule_index))
+            # assert(isinstance(resource_ids, list))
+            rule = Rule(rule_name=rule_def.get('name'),
+                        rule_index=rule_index,
+                        min_retention=min_retention,
+                        max_retention=max_retention)
+
+            for rid in resource_ids:
+                if rid == '*':
+                    raise audit_errors.InvalidRulesSchemaError(
+                        'The symbol * is not allowed in rule {}'.format(
+                            rule_index))
+
+                gcp_resource = resource_util.create_resource(
+                    resource_id=rid,
+                    resource_type=resource_type)
+
+                self.resource_rules_map[apply_to][gcp_resource].add(
+                    rule)
+
+    def get_resource_rules(self, applies_to):
+        """Get all the resource rules for (resource, RuleAppliesTo.*).
+        Args:
+            applies_to (str): The name of applies_to resource
+
+        Returns:
+           defaultdict:  A list of ResourceRules.
+        """
+        return self.resource_rules_map[applies_to]
+
+
+class Rule(object):
+    """Rule properties from the rule definition file.
+    Also finds violations.
+    """
+
+    RuleViolation = namedtuple(
+        'RuleViolation',
+        ['resource_name', 'resource_type', 'full_name', 'rule_name',
+         'rule_index', 'violation_type', 'violation_describe'])
+
+    def __init__(self, rule_name, rule_index, min_retention, max_retention):
+        """Initialize.
+
+        Args:
+            rule_name (str): Name of the loaded rule
+            rule_index (int): The index of the rule
+            min_retention(int): minimum value of the age in lifecycle
+            max_retention(int): maximum value of the age in lifecycle
+        """
+        self.rule_name = rule_name
+        self.rule_index = rule_index
+        self.min_retention = min_retention
+        self.max_retention = max_retention
+
+    def generate_rule_violation(self, buckets_lifecycle, describe):
+        """generate a violation.
+
+        Args:
+            buckets_lifecycle (RetentionBucket): The info of the bucket
+            describe (str): The description of the violation
+        Returns:
+            RuleViolation: The violation
+        """
+
+        return self.RuleViolation(
+            resource_name=buckets_lifecycle.id,
+            resource_type=buckets_lifecycle.type,
+            full_name=buckets_lifecycle.full_name,
+            rule_name=self.rule_name,
+            rule_index=self.rule_index,
+            violation_type=VIOLATION_TYPE,
+            violation_describe=describe
+        )
+
+    # TODO: The naming is confusing and needs to be fixed in all scanners.
+    def find_buckets_violations(self, buckets_lifecycle):
```

Comment:
Again, perhaps we can use just a single function to work for all resources. Else there will be a lot of unnecessary duplication.